### PR TITLE
feat: add ft_dprintf implementation and related functions

### PR DESCRIPTION
**Pull request by Extraden**
This pull request introduces a new `ft_dprintf` functionality, which is a file descriptor-based variant of `ft_printf`. It includes updates to the `Makefile` to integrate the new module and adds several new files implementing the `ft_dprintf` functionality. Below are the key changes grouped by theme:

### Build System Updates
* Added `DPRINTF_DIR` to the `Makefile` to include the new `ft_dprintf` directory.
* Defined `FT_DPRINTF_SRCS` in the `Makefile` to specify the source files related to `ft_dprintf`.
* Updated the `SRCS` variable in the `Makefile` to include `FT_DPRINTF_SRCS`.
* Added `ft_dprintf.h` to the `HEADERS` list in the `Makefile`.

### New `ft_dprintf` Implementation
* Introduced the `ft_dprintf.h` header, which declares the main `ft_dprintf` function and its helper functions.
* Added the main implementation of `ft_dprintf` in `ft_dprintf.c`, including parsing and processing logic for various format specifiers.

### Supporting Functions for `ft_dprintf`
* Implemented number-related helper functions in `ft_number_functions.c`, such as `ft_dputnbr`, `ft_unsigned_dputnbr`, and `ft_dputhex`.
* Added pointer formatting support in `ft_ptr_functions.c` with the `ft_dputptr` function.
* Implemented string and character output functions in `ft_string_functions.c`, including `ft_putchar_fd` and `ft_dputstr`.

### CODE REVIEW

**Review by tigran-sargsyan-w — COMMENTED** [review]
diff --git a/src/ft_dprintf/ft_number_functions.c b/src/ft_dprintf/ft_number_functions.c
index 7b2aac5..e6eaac4 100644
--- a/src/ft_dprintf/ft_number_functions.c
+++ b/src/ft_dprintf/ft_number_functions.c
@@ -3,10 +3,10 @@
 /*                                                        :::      ::::::::   */
 /*   ft_number_functions.c                              :+:      :+:    :+:   */
 /*                                                    +:+ +:+         +:+     */
-/*   By: dsemenov <dsemenov@student.42.fr>          +#+  +:+       +#+        */
+/*   By: tsargsya <tsargsya@student.42.fr>          +#+  +:+       +#+        */
 /*                                                +#+#+#+#+#+   +#+           */
 /*   Created: 2024/12/18 14:53:38 by tsargsya          #+#    #+#             */
-/*   Updated: 2025/05/18 19:16:40 by dsemenov         ###   ########.fr       */
+/*   Updated: 2025/05/20 12:14:35 by tsargsya         ###   ########.fr       */
 /*                                                                            */
 /* ************************************************************************** */
 
@@ -87,7 +87,7 @@ ssize_t       ft_unsigned_dputnbr(int fd, unsigned int nbr)
                        return (-1);
                len += res;
        }
-       if (ft_putchar_fd(fd, ((nbr % 10) + '0') == -1))
+       if (ft_putchar_fd(fd, (nbr % 10) + '0') == -1)
                return (-1);
        len += 1;
        return (len);

**Review by tigran-sargsyan-w — COMMENTED** [review]
➜  libft git:(pr/Extraden/1) ✗ cc main.c -lft
In file included from main.c:3:
ft_dprintf.h:19:25: error: conflicting types for ‘ft_putchar_fd’; have ‘ssize_t(int,  char)’ {aka ‘long int(int,  char)’}
   19 | ssize_t                 ft_putchar_fd(int fd, char c);
      |                         ^~~~~~~~~~~~~
In file included from main.c:1:
libft.h:41:33: note: previous declaration of ‘ft_putchar_fd’ with type ‘void(char,  int)’
   41 | void                            ft_putchar_fd(char c, int fd);

**Review by tigran-sargsyan-w — COMMENTED** [review]
```
➜  libft git:(pr/Extraden/1) ✗ cc main.c -lft
In file included from main.c:3:
ft_dprintf.h:19:25: error: conflicting types for ‘ft_putchar_fd’; have ‘ssize_t(int,  char)’ {aka ‘long int(int,  char)’}
   19 | ssize_t                 ft_putchar_fd(int fd, char c);
      |                         ^~~~~~~~~~~~~
In file included from main.c:1:
libft.h:41:33: note: previous declaration of ‘ft_putchar_fd’ with type ‘void(char,  int)’
   41 | void                            ft_putchar_fd(char c, int fd);
```

**Review by tigran-sargsyan-w — COMMENTED** [review]
Summary
Solid ft_dprintf addition and Makefile wiring. The split into number/string/ptr helpers is clear. A few naming/API and edge-case issues to address.

Notes

Naming/API clash with libft

ft_dprintf.h declares ft_putchar_fd(int fd, char c) returning ssize_t, which conflicts with classic libft void ft_putchar_fd(char c, int fd) (param order + return type). Please namespace the fd-aware variants (e.g. ft_dputchar, ft_dputstr) to avoid symbol and header collisions. Also keep param order consistent. 
[GitHub](https://github.com/tigran-sargsyan-w/Libft_Extended/pull/1)

Return-type consistency

Some helpers return ssize_t but then increment len manually per character; ensure the contract is: -1 on write error, otherwise total bytes written. Double-check all early returns propagate -1 (e.g., after nested calls). 
[GitHub](https://github.com/tigran-sargsyan-w/Libft_Extended/pull/1)

Edge cases

Numbers: test 0, negative values, and INT_MIN for decimal; for hex, both lower/upper, and 0 with/without prefix.

Pointers: decide on (nil) vs 0x0 formatting consistency; ensure NULL prints predictably.

Strings: handle NULL as "(null)" (or documented alternative), and verify length accounting matches bytes written. 
[GitHub](https://github.com/tigran-sargsyan-w/Libft_Extended/pull/1)

I/O robustness

All writes use single write() calls. Consider a tiny wrapper that retries on EINTR and loops until all bytes are written to avoid partial writes on non-regular fds. (Even if overkill for libft, it’s a safer primitive.)

Header/visibility

Keep only the public API in ft_dprintf.h; mark internal helpers static in their .c files or move to a private header to reduce symbol leakage.

Makefile

Nice modular inclusion via DPRINTF_DIR. Consider grouping objects into an archive section or adding a TEST_DPRINTF target to run a few golden-case checks (stdout vs fd redirection). 
[GitHub](https://github.com/tigran-sargsyan-w/Libft_Extended/pull/1)

Decision: Comment (non-blocking).
If the ft_putchar_fd naming/API clash is fixed and tests added for the edge cases above, I’m happy to Approve.
